// nesh/lib/plugins/version.js
// Generated by CoffeeScript 1.7.1

/*
Version Information Plugin
==========================
This plugin adds a new command to the repl called `.versions` that
will print out version information for nesh, any loaded languages,
and Node itself. The new command is listed with a description when
you run `.help`.
 */

(function() {
  var nesh;

  nesh = require('../nesh');

  exports.name = 'version';

  exports.description = 'Adds a .versions command';

  exports.setup = function(context) {
    return process.versions.nesh = nesh.version;
  };

  exports.postStart = function(context) {
    var repl;
    repl = context.repl;
    return repl.commands['.versions'] = {
      help: 'Show Node version information',
      action: function() {
        var key, value, versions;
        versions = ((function() {
          var _ref, _results;
          _ref = process.versions;
          _results = [];
          for (key in _ref) {
            value = _ref[key];
            _results.push("" + key + " " + value);
          }
          return _results;
        })()).join('\n');
        repl.outputStream.write("" + versions + "\n");
        return repl.displayPrompt();
      }
    };
  };

}).call(this);
